source/common/http/http2/codec_impl.cc

```diff
@@ -900,51 +905,78 @@ void ConnectionImpl::sendPendingFrames() {
+    if (it.identifier().value() == NGHTTP2_SETTINGS_ENABLE_PUSH && it.value().value() == 1) {
+      // For simplicity, warn when server push is configured enabled on either client or server
+      // connections. It _is_ actually enabled on downstream connections although Envoy itself will
+      // never send push_promise frames, but the nuance could be confusing to end users.
+      ENVOY_CONN_LOG(
+          warn,
+          "server push is not supported by Envoy and can not be enabled via a SETTINGS parameter.",
+          connection_);
+      continue;
+    }
+    auto result =
+        settings.insert({static_cast<int32_t>(it.identifier().value()), it.value().value()});
+    if (!result.second) {
+      ENVOY_CONN_LOG(debug, "duplicate user defined settings parameter with id {:#x}, value {}",
+                     connection_, it.identifier().value(), it.value().value());
+      continue;
+    }
+    ENVOY_CONN_LOG(debug, "adding user defined settings parameter with id {:#x} to {}", connection_,
+                   it.identifier().value(), it.value().value());
+  }
+
+  struct SettingsDescriptor {
+    nghttp2_settings_entry entry;
+    uint32_t default_value;
+  };
+  auto descriptors = std::vector<SettingsDescriptor>{
+      {{/*nghttp2_settings_entry.settings_id=*/NGHTTP2_SETTINGS_ENABLE_CONNECT_PROTOCOL,
+        /*nghttp2_settings_entry.value=*/http2_options.allow_connect()},
+       /*default_value=*/0},
+      {{NGHTTP2_SETTINGS_HEADER_TABLE_SIZE, http2_options.hpack_table_size().value()},
+       NGHTTP2_DEFAULT_HEADER_TABLE_SIZE},
+      {{NGHTTP2_SETTINGS_MAX_CONCURRENT_STREAMS, http2_options.max_concurrent_streams().value()},
+       NGHTTP2_INITIAL_MAX_CONCURRENT_STREAMS},
+      {{NGHTTP2_SETTINGS_INITIAL_WINDOW_SIZE, http2_options.initial_stream_window_size().value()},
+       NGHTTP2_INITIAL_WINDOW_SIZE}};
+  for (const auto& descriptor : descriptors) {
+    if (descriptor.entry.value != descriptor.default_value) {
+      auto result = settings.insert(descriptor.entry);
+      // Config validation should ensure a collision with user defined parameters is not possible.
+      // There is one exception, `allow_connect`, which requires a breaking API change to enable
+      // presence checks during validation.
+      if (!result.second) {
+        ASSERT(descriptor.entry.settings_id == NGHTTP2_SETTINGS_ENABLE_CONNECT_PROTOCOL);
+        ENVOY_CONN_LOG(debug,
+                       "overriding named settings parameter with id {:#x}, value {} due to "
+                       "collision with user defined parameter",
+                       connection_, descriptor.entry.settings_id, descriptor.entry.value);
+        continue;
+      }
+      ENVOY_CONN_LOG(debug, "adding named settings parameter with id {:#x} to {}", connection_,
+                     descriptor.entry.settings_id, descriptor.entry.value);
+    }
   }
 
-  if (!iv.empty()) {
+  if (!settings.empty()) {
+    std::vector<nghttp2_settings_entry> iv;
```

Comment:
InlinedVector here as well with a reserve?

Reply:
This is now addressed with a single InlinedVector.